File: Medium/5. Longest Palindromic Substring/longestPalindrome.cpp
/*
    參考:https://zxi.mytechroad.com/blog/dynamic-programming/leetcode-5-longest-palindromic-substring/
    time complexity: O(n^2)
    space complexity: O(1)
*/
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        auto getLen = [&](int l, int r){
            while(l >= 0 && r < n && s[l] == s[r]){
                l--;
                r++;
            }
            return r - l - 1;
        };
        int len = 0;
        int start = 0;
        for(int i = 0; i < n; i++){
            int cur = max(getLen(i, i), getLen(i, i + 1));
            if(cur > len){
                len = cur;
                start = i - (len - 1) / 2;
            }
        }
        return s.substr(start, len);
    }
};

File: Medium/5. Longest Palindromic Substring/longestPalindrome.java

class Solution {
    int start, offset;    
    public String longestPalindrome(String s) {
        String ans = "";
        for(int i = 0; i < s.length(); i++){
            pal(s, i, i); //找到以s[i]為中心的回文串
            pal(s, i, i+1); //找到以s[i]為中心的回文串
        }
        return s.substring(start, start + offset);
    }
    private void pal(String s, int l, int r){
        while(l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)){
            l--;
            r++;
        }
        if(offset < r - l - 1){ //如果前一次儲存的位移小於現在的位移值(也就是找到比較長的回文串)，就更新值
            start = l + 1;
            offset = r - l - 1;
        }
    }
}